### PR TITLE
Feature: TRN-55 Update User Interests

### PR DESCRIPTION
**Pull request by yusufnasserdev**
# Add Update Interests Feature

## Description
This PR introduces a new "Update Interests" feature that allows users to modify their selected categories/interests after initial selection. The implementation includes both backend API support and frontend UI components.

### Backend Integration
- Added `PATCH` method support to `NetworkClient` for partial updates
- Created new DTOs for patch requests/responses:
  - `PatchUserInterestsRequest` - handles categories to add/remove
  - `PatchUserInterestsResponse` - returns updated categories
- Enhanced `CategoryDto` with `isSelected` field to track user selections
- Updated `CategoryRepository` with `patchUserCategories()` method that calculates differences between original and current selections

### UI Screen
- New "Update Interests" screen accessible from main container
- Interactive category selection with visual feedback
- Save functionality that only sends changed categories (add/remove)
- Proper loading states and error handling
- Navigation integration with new `UpdateInterests` route

### Business Logic
- Smart diff calculation between original and current selections
- Only sends categories that were added or removed
- Maintains selection state across screen interactions
- Validation to enable save button only when changes exist

## Testing
- Unit tests for category mapping with selection state
- Integration tests for patch request payloads
- UI tests for interaction flows
- Edge cases: empty changes, only additions, only removals

This feature provides users with flexibility to update their interests over time while minimizing API payload size by only sending changes.

### CODE REVIEW

File: trends_data/src/commonMain/kotlin/net/thechance/mena/trends/data/client/NetworkClient.kt [hunk 1]
```diff
@@ -20,6 +20,11 @@ interface NetworkClient {
         block: HttpRequestBuilder.() -> Unit = {}
     ): HttpResponse
 
+    suspend fun patch(
+        urlString: String,
```
Comment (Shahd-Hatem) [inline trends_data/src/commonMain/kotlin/net/thechance/mena/trends/data/client/NetworkClient.kt:24]: Why urlString?
I recommend that Just rename it url

Comment (yusufnasserdev) [reply trends_data/src/commonMain/kotlin/net/thechance/mena/trends/data/client/NetworkClient.kt:24]: `urlString` is more descriptive than just `url` as a `URL` can be its own entity given its components, also this is the same convention followed by the `NetworkClient` interface and its implementation.

File: trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/main_container/MainContainerScreen.kt [hunk 2]
```diff
@@ -91,6 +97,17 @@ private fun MainContainerScreenContent(
...
             }
+
+            Row(
+                modifier = Modifier.fillMaxWidth(),
+                horizontalArrangement = Arrangement.SpaceEvenly
+            ) {
+                Button(
+                    onClick = onClickUpdateInterests,
+                ) {
+                    Text("Update Interests", style = Theme.typography.title.medium)
```
Comment (Shahd-Hatem) [inline trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/main_container/MainContainerScreen.kt:108]: Put this hard coded string at string resources

File: trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/update_categories/UpdateCategoriesViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    init {
+        getCategories()
+    }
+
+    fun getCategories() {
+        tryToExecute(
+            block = { repository.getAllCategories() },
+            onSuccess = ::handleLoadCategoriesSuccess,
+            onError = { errorState -> updateState { copy(errorState = errorState) } },
+            onStart = { updateState { copy(isLoading = true) } },
```
Comment (Shahd-Hatem) [inline trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/update_categories/UpdateCategoriesViewModel.kt:32]: I think don’t need to update state of loading here because already loading has a default value == true

File: trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/main_container/MainContainerScreen.kt (hunk 2 above)
Comment (yusufnasserdev) [reply trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/main_container/MainContainerScreen.kt:108]: This is just a temporary button till the main screen is integrated, should be removed soon.

**Review by Shahd-Hatem — APPROVED** [review]

**Comment by Shahd-Hatem** [comment]
Good work yussef👏

File: trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/update_categories/UpdateCategoriesViewModel.kt (hunk 3 above)
Comment (yusufnasserdev) [reply trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/update_categories/UpdateCategoriesViewModel.kt:32]: I understand your point, however, the state can be changed during the runtime after the categories are loaded initially to `false`, then if the user were to navigate back, we will need to update it to `true` while re-fetching them.

File: trends_data/src/commonMain/kotlin/net/thechance/mena/trends/data/dto/PatchUserInterestsRequest.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package net.thechance.mena.trends.data.dto
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+internal data class PatchUserInterestsRequest(
```
Comment (iZyadNasser) [inline trends_data/src/commonMain/kotlin/net/thechance/mena/trends/data/dto/PatchUserInterestsRequest.kt]: rename any interest to category for consistency in the project

File: trends_data/src/commonMain/kotlin/net/thechance/mena/trends/data/repository/CategoryRepositoryImpl.kt [hunk 5]
```diff
@@ -42,4 +44,24 @@ internal class CategoryRepositoryImpl(
...
+    override suspend fun patchUserCategories(
+        originalSelectedIds: List<String>,
+        currentSelectedIds: List<String>
+    ) {
+        val toAdd = currentSelectedIds.filterNot { it in originalSelectedIds }
+        val toRemove = originalSelectedIds.filterNot { it in currentSelectedIds }
+
+        safeApiCall<PatchUserInterestsResponse> {
+            networkClient.patch("/$TRENDS_PATH/$CATEGORIES_ENDPOINT") {
+                contentType(io.ktor.http.ContentType.Application.Json)
```
Comment (iZyadNasser) [inline trends_data/src/commonMain/kotlin/net/thechance/mena/trends/data/repository/CategoryRepositoryImpl.kt]: why not import this instead of writing the whole path

File: trends_data/src/commonTest/kotlin/net/thechance/mena/trends/data/repository/util/engine.kt [hunk 6]
```diff
@@ -72,6 +73,11 @@ internal fun createReelsHttpClient(
             block: HttpRequestBuilder.() -> Unit
         ): HttpResponse = client.put(urlString, block)
 
+        override suspend fun patch(
+            urlString: String,
+            block: HttpRequestBuilder.() -> Unit
+        ): HttpResponse = client.patch(urlString, block)
```
Comment (iZyadNasser) [inline trends_data/src/commonTest/kotlin/net/thechance/mena/trends/data/repository/util/engine.kt:79]: you have one function that is compact and another not compact, choose one approach for consistency

Comment (yusufnasserdev) [reply trends_data/src/commonTest/kotlin/net/thechance/mena/trends/data/repository/util/engine.kt:79]: I'm following the convention of this file, as the other 4 functions.

**Review by iZyadNasser — APPROVED** [review]

**Review by MahmoodTarek — APPROVED** [review]

File: trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/update_categories/UpdateCategoriesScreen.kt [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    CategoryPickScreenContent(
+        state = state,
+        listener = viewModel
+    )
+}
+
+@Composable
+private fun CategoryPickScreenContent(
```
Comment (HendSayed25) [inline trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/update_categories/UpdateCategoriesScreen.kt]: I think this name need to changed to 
```suggestion
private fun UpdateCategoryScreenContent(
```

Comment (yusufnasserdev) [reply trends_presentation/src/commonMain/kotlin/net/thechance/mena/trends/presentation/screen/update_categories/UpdateCategoriesScreen.kt]: Done.

**Review by HendSayed25 — APPROVED** [review]
Good job 👏
